diff --git a/libsuspend/Android.bp b/libsuspend/Android.bp
index c5f1f5e..0f835b0 100644
--- a/libsuspend/Android.bp
+++ b/libsuspend/Android.bp
@@ -4,6 +4,7 @@ cc_library {
     name: "libsuspend",
     srcs: [
         "autosuspend.c",
+        "autosuspend_earlysuspend.cpp",
         "autosuspend_wakeup_count.cpp",
     ],
     export_include_dirs: ["include"],
diff --git a/libsuspend/autosuspend.c b/libsuspend/autosuspend.c
index b87f59c..607025c 100644
--- a/libsuspend/autosuspend.c
+++ b/libsuspend/autosuspend.c
@@ -32,12 +32,22 @@ static int autosuspend_init(void) {
         return 0;
     }
 
+    autosuspend_ops = autosuspend_earlysuspend_init();
+    if (autosuspend_ops) {
+        goto out;
+    }
+
     autosuspend_ops = autosuspend_wakeup_count_init();
-    if (autosuspend_ops == NULL) {
+    if (autosuspend_ops) {
+        goto out;
+    }
+
+    if (!autosuspend_ops) {
         ALOGE("failed to initialize autosuspend");
         return -1;
     }
 
+out:
     ALOGV("autosuspend initialized");
     return 0;
 }
diff --git a/libsuspend/autosuspend_ops.h b/libsuspend/autosuspend_ops.h
index b0024c8..a21f207 100644
--- a/libsuspend/autosuspend_ops.h
+++ b/libsuspend/autosuspend_ops.h
@@ -25,6 +25,7 @@ struct autosuspend_ops {
 };
 
 __BEGIN_DECLS
+struct autosuspend_ops *autosuspend_earlysuspend_init(void);
 struct autosuspend_ops *autosuspend_wakeup_count_init(void);
 __END_DECLS
 
