diff --git a/media/libstagefright/MPEG4Writer.cpp b/media/libstagefright/MPEG4Writer.cpp
index f130c9b..7f791f1 100644
--- a/media/libstagefright/MPEG4Writer.cpp
+++ b/media/libstagefright/MPEG4Writer.cpp
@@ -41,6 +41,10 @@
 #include <media/stagefright/foundation/avc_utils.h>
 #include <media/stagefright/MPEG4Writer.h>
 #include <media/stagefright/MediaBuffer.h>
+
+#ifndef MTK_HARDWARE
+#define MTK_HARDWARE
+#endif
 #include <media/stagefright/MetaData.h>
 #include <media/stagefright/MediaDefs.h>
 #include <media/stagefright/MediaErrors.h>
@@ -1403,6 +1407,12 @@ static void StripStartcode(MediaBuffer *buffer) {
         buffer->set_range(
                 buffer->range_offset() + 4, buffer->range_length() - 4);
     }
+#ifdef MTK_HARDWARE //&& defined(MTK_VIDEO_HEVC_SUPPORT)
+    else if (!memcmp(ptr, "\x00\x00\x01", 3)) {
+        ALOGV("StripStartcode 00 00 01 for HEVC directlink");
+        buffer->set_range(buffer->range_offset() + 3, buffer->range_length() - 3);
+    }
+#endif
 }
 
 void MPEG4Writer::addMultipleLengthPrefixedSamples_l(MediaBuffer *buffer) {
