diff --git a/media/libmedia/MediaProfiles.cpp b/media/libmedia/MediaProfiles.cpp
index 366d90d..12bb7b5 100644
--- a/media/libmedia/MediaProfiles.cpp
+++ b/media/libmedia/MediaProfiles.cpp
@@ -29,6 +29,10 @@
 #include <OMX_Video.h>
 #include <sys/stat.h>
 
+#ifndef MTK_HARDWARE
+#define MTK_HARDWARE
+#endif
+
 namespace android {
 
 constexpr char const * const MediaProfiles::xmlFiles[];
@@ -40,6 +44,9 @@ const MediaProfiles::NameToTagMap MediaProfiles::sVideoEncoderNameMap[] = {
     {"h263", VIDEO_ENCODER_H263},
     {"h264", VIDEO_ENCODER_H264},
     {"m4v",  VIDEO_ENCODER_MPEG_4_SP},
+#ifdef MTK_HARDWARE
+    {"h265", VIDEO_ENCODER_H265},
+#endif
     {"hevc", VIDEO_ENCODER_HEVC}
 };
 
@@ -50,6 +57,10 @@ const MediaProfiles::NameToTagMap MediaProfiles::sAudioEncoderNameMap[] = {
     {"heaac",  AUDIO_ENCODER_HE_AAC},
     {"aaceld", AUDIO_ENCODER_AAC_ELD}, 
     {"opus",   AUDIO_ENCODER_OPUS}
+#ifdef MTK_HARDWARE
+    {"vorbis", AUDIO_ENCODER_VORBIS},
+    {"lpcm",  AUDIO_ENCODER_LPCM},
+#endif
 };
 
 const MediaProfiles::NameToTagMap MediaProfiles::sFileFormatMap[] = {
diff --git a/media/libmedia/include/media/MediaProfiles.h b/media/libmedia/include/media/MediaProfiles.h
index 3e8e7c8..bcb7f5c 100644
--- a/media/libmedia/include/media/MediaProfiles.h
+++ b/media/libmedia/include/media/MediaProfiles.h
@@ -21,6 +21,10 @@
 #include <utils/threads.h>
 #include <media/mediarecorder.h>
 
+#ifndef MTK_HARDWARE
+#define MTK_HARDWARE
+#endif
+
 namespace android {
 
 enum camcorder_quality {
diff --git a/media/libmedia/include/media/mediarecorder.h b/media/libmedia/include/media/mediarecorder.h
index 2dd4b7f..bcf1d59 100644
--- a/media/libmedia/include/media/mediarecorder.h
+++ b/media/libmedia/include/media/mediarecorder.h
@@ -26,6 +26,10 @@
 #include <media/IMediaDeathNotifier.h>
 #include <media/MicrophoneInfo.h>
 
+#ifndef MTK_HARDWARE
+#define MTK_HARDWARE
+#endif
+
 namespace android {
 
 class Surface;
@@ -108,7 +112,14 @@ enum video_encoder {
     VIDEO_ENCODER_VP8 = 4,
     VIDEO_ENCODER_HEVC = 5,
 
+#ifdef MTK_HARDWARE
+    VIDEO_ENCODER_LIST_END, // must be the last - used to validate the video encoder type
+    VIDEO_ENCODER_LIST_VENDOR_START = 1000,
+    VIDEO_ENCODER_H265 = 1001,
+    VIDEO_ENCODER_LIST_VENDOR_END
+#else
     VIDEO_ENCODER_LIST_END // must be the last - used to validate the video encoder type
+#endif
 };
 
 /*
