diff --git a/media/libstagefright/ACodec.cpp b/media/libstagefright/ACodec.cpp
index 7948297..979198a 100644
--- a/media/libstagefright/ACodec.cpp
+++ b/media/libstagefright/ACodec.cpp
@@ -3449,7 +3449,7 @@ status_t ACodec::setSupportedOutputFormat(bool getLegacyFlexibleFormat) {
                 || format.eColorFormat == OMX_TI_COLOR_FormatYUV420PackedSemiPlanar
 #ifdef MTK_HARDWARE
-                || format.eColorFormat == HAL_PIXEL_FORMAT_YV12
+                || (int32_t)format.eColorFormat == (int32_t)HAL_PIXEL_FORMAT_YV12
                 || format.eColorFormat == OMX_MTK_COLOR_FormatYV12
 #endif
                 ) {
