diff --git a/headers/media_plugin/media/openmax/OMX_IVCommon.h b/headers/media_plugin/media/openmax/OMX_IVCommon.h
index f83114b..2eab2e3 100644
--- a/headers/media_plugin/media/openmax/OMX_IVCommon.h
+++ b/headers/media_plugin/media/openmax/OMX_IVCommon.h
@@ -148,6 +148,11 @@ typedef enum OMX_COLOR_FORMATTYPE {
     OMX_COLOR_Format24BitARGB6666,
     OMX_COLOR_Format24BitABGR6666,
     OMX_COLOR_FormatKhronosExtensions = 0x6F000000, /**< Reserved region for introducing Khronos Standard Extensions */
+    /* MTK proprietary formats */
+    OMX_MTK_COLOR_FormatYV12 = 0x7F000001,
+    OMX_COLOR_FormatVendorMTKYUV = 0x7F000002,
+    OMX_COLOR_FormatVendorMTKYUV_FCM = 0x7F000003,
+
     OMX_COLOR_FormatVendorStartUnused = 0x7F000000, /**< Reserved region for introducing Vendor Extensions */
     /**<Reserved android opaque colorformat. Tells the encoder that
      * the actual colorformat will be  relayed by the
