From 80f09c92fa803acad772fa08ed79c2b78873540b Mon Sep 17 00:00:00 2001
From: Bilux <i.bilux@gmail.com>
Date: Wed, 27 Jan 2021 17:57:43 +0100
Subject: [PATCH] Enable fallback to old ro.sf.hwrotation property

diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -447,6 +447,21 @@ SurfaceFlinger::SurfaceFlinger(Factory& factory) : SurfaceFlinger(factory, SkipI
     }
     ALOGV("Internal Display Orientation: %s", toCString(internalDisplayOrientation));

+    if (internalDisplayOrientation == ui::ROTATION_0) {
+        int sfRotation = property_get_int32("ro.sf.hwrotation", -1);
+        switch(sfRotation) {
+            case 90:
+                internalDisplayOrientation = ui::ROTATION_90;
+                break;
+            case 180:
+                internalDisplayOrientation = ui::ROTATION_180;
+                break;
+            case 270:
+                internalDisplayOrientation = ui::ROTATION_270;
+                break;
+        }
+    }
+
     mInternalDisplayPrimaries = sysprop::getDisplayNativePrimaries();
     if (!mInternalDisplayPrimaries) {
         ALOGW("Internal display primitives are not set");
